packages/actor-optimize-query-operation-construct-distinct/lib/ActorOptimizeQueryOperationConstructDistinct.ts

```diff
@@ -0,0 +1,38 @@
+import type {
+  IActionOptimizeQueryOperation,
+  IActorOptimizeQueryOperationOutput,
+  IActorOptimizeQueryOperationArgs,
+} from '@comunica/bus-optimize-query-operation';
+import { ActorOptimizeQueryOperation } from '@comunica/bus-optimize-query-operation';
+import { KeysInitQuery } from '@comunica/context-entries/lib/Keys';
+import type { IActorTest } from '@comunica/core';
+import type { Algebra, Factory } from 'sparqlalgebrajs';
+import { Util } from 'sparqlalgebrajs';
+
+/**
+ * A comunica Construct Distinct Optimize Query Operation Actor.
+ */
+export class ActorOptimizeQueryOperationConstructDistinct extends ActorOptimizeQueryOperation {
+  public constructor(args: IActorOptimizeQueryOperationArgs) {
+    super(args);
+  }
+
+  public async test(action: IActionOptimizeQueryOperation): Promise<IActorTest> {
+    if (!action.context.has(KeysInitQuery.distinct)) {
+      throw new Error(`${this.name} was not enabled by the query.`);
+    }
+    return true;
+  }
+
+  public async run(action: IActionOptimizeQueryOperation): Promise<IActorOptimizeQueryOperationOutput> {
+    const operation = Util.mapOperation(action.operation, {
+      construct(op: Algebra.Construct, factory: Factory) {
+        return {
+          recurse: false,
+          result: factory.createDistinct(factory.createConstruct(op.input, op.template)),
+        };
+      },
+    });
+    return { operation, context: action.context };
```

Comment:
Maybe we'll want to remove `KeysInitQuery.distinct` from the context at this stage.

Reply:
We could do so if we're sure this context will only be used for one query.
I'm not sure how it works so is there the possibility that someone creates a query engine with a context containing the `distinct` key, where if we delete that entry it influences later queries?

Reply:
You just created this `distinct` key for a specific purpose, so we're certain there's nothing else using this key for something else.

Reply:
Yes but I mean that if I create a query engine with `distinct´ in its context and my first query deletes that context entry, I may still expect that entry to be there for subsequent queries is what I mean and then I might spend some time looking for why subsequent queries are not being deduplicated. Anyways I've changed it to delete this key.

Reply:
Contexts are immutable, so that won't be an issue.